### ISSUE TITLE
Analyzer contains redundant (and divergent) implementations of subtype testing

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#34653]
The analyzer currently has three separate implementations of `isSubtypeOf` and `isMoreSpecificThan`:
- An implementation in the deprecated class `TypeSystemImpl` (which presumably implements Dart 1.0 semantics)
- An implementation in the class `StrongTypeSystemImpl` (which AFAICT is the most up to date)
- An implementation in the `DartType` class hierarchy (which I don't know the status of)

Additionally, I suspect (but have not yet confirmed with the language team) that as of Dart 2.0, there is no longer any meaningful distinction between `isSubtypeOf` and `isMoreSpecificThan`.  If that's the case, then we have 6 (!) different implementations of the same core algorithm.

This is a major footgun for analyzer clients, since it's not immediately obvious which implementation ought to be used.  This is at the root of #34641; the linter is using `DartType.isMoreSpecificThan` to determine whether two types are related for the purposes of the `unrelated_type_equality_checks` lint (see `DartTypeUtilities.unrelatedTypes`), but this method was not updated to handle mixins.

I would like to see us deprecate all but one of these subtype testing methods and then remove them the next time a breaking change to the analyzer is published.

**Comment by leafpetersen** [comment]
> Additionally, I suspect (but have not yet confirmed with the language team) that as of Dart 2.0, there is no longer any meaningful distinction between `isSubtypeOf` and `isMoreSpecificThan`. I

Correct.  From type_system.dart:

```
  @override
  bool isMoreSpecificThan(DartType t1, DartType t2) => isSubtypeOf(t1, t2);
```

**Comment by bwilkerson** [comment]
Note that the spec has not yet been updated to reflect that.

**Comment by leafpetersen** [comment]
cc @eernstg we should put this on the TODO list (probably after integrating the feature specs).

**Comment by eernstg** [comment]
> Note that the spec has not yet been updated to reflect that

Right, I consider the integration of [subtyping.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/subtyping.md) into the language specification to contain the elimination of the 'more specific than' relationship from the language specification as a subtask (as indicated [here](https://docs.google.com/spreadsheets/d/1zRJMJxgvKLWC1PyuNCCxLq8NBUZW1L2e1rHkWkeEc_c/edit#gid=0&range=D13)).

**Comment by stereotype441** [comment]
Based on discussion with @bwilkerson this morning, we've decided to introduce a new API for clients to use for subtype and assignability testing (see #35346).  The new API will be similar to `TypeSystem` but with less functionality exposed.  Assuming we can get that API added quickly, we'll transition clients to it and deprecate all other subtype and assignability testing APIs in favor of it.

(If adding the new API takes too long, we'll transition clients to `TypeSystem` as a temporary measure, so that we no longer have bugs due to the divergent implementations.)

I'm re-purposing this bug to track that work.

**Comment by eernstg** [comment]
I did notice that @stereotype441 wrote:

> I'm re-purposing this bug to track that work.

but it might still be relevant to follow up on [this comment](https://github.com/dart-lang/sdk/issues/34653#issuecomment-426656422):

> integration of subtyping.md into the language specification

That integration has now been landed as commit 00cec998f5ebde1dc9787663b3d6b04e83371d39, and the language specification does not contain any references to the 'more specific than' type relation any more, nor the associated operator `<<`.

So there's no reason to keep anything related to the 'more specific than' relation around in any tool, except for resource considerations about the work needed to take it out.

**Comment by stereotype441** [comment]
Unassigning myself because I’m not actively working on this issue right now.
